### DISCUSSION TITLE
Transmission formulaire terminée avec erreur

### DISCUSSION THREAD

**Discussion by atifmohamed** [MTESSDev/FRW#79] [Annonces]
Des formulaires ont été transmis au MESS mais apparemment la transmission a terminé avec erreur car il y avait la tâche generePdf en trop dans le fichier default.v0.transmission.yml
Le fichier default.v0.transmission.yml est maintenant corrigé. 
Un autre formulaire a été transmis en date du 2023-05-04 à 09:12:17 (numéro de confirmation 42816311).

Pourriez-vous vérifier si le formulaire a été reçu avec succès par le MESS et que sa retransmission à RQ est effectuée?  

Y a-t-il un moyen qui nous permet de vérifier l’état de transmission d’un formulaire sans passer votre équipe ?

**Comment by loima02** [comment]
> Y a-t-il un moyen qui nous permet de vérifier l’état de transmission d’un formulaire sans passer votre équipe ?

Oui. À partir de la page `Démarrage d'essais`
- activer la case à cocher `Afficher les outils de développement`
- ouvrir le formulaire à tester
- cliquer sur le bouton `Tester transmission` dans la section `Outils de développement`
- développer les `Messages des tâches`
- développer `AppelerServiceExterne->ReponseServer_appel_externe`

Pour plus d'info sur les outils de développement, consulter cette page [Documentation/outils-developpement.md](Documentation/outils-developpement.md)

Dans votre exemple, le message d'erreur affiché est :
> The URL you requested has been blocked
FortiGuard Intrusion Prevention - Access Blocked
Web Page Blocked
The page you have requested has been blocked because the URL is banned.

Mon serveur ne peut pas appeler votre serveur en raison de l'infra. Pourriez-vous m'envoyer par courriel les adresses IP à débloquer dans nos batillons s.v.p? Je vais faire une demande au MCN.

**Comment by atifmohamed** [comment]
Merci pour la réponse, on vous revient très prochainement sur les adresses IP de nos serveurs.
Merci encore une fois.

**Comment by atifmohamed** [comment]
Est-ce que je peux savoir sur quelle adresse courriel nous vous envoyons l'adresse IP de notre serveur ?

**Comment by loima02** [comment]
Bonjour!

Le MCN m'informe ce matin que la demande est réalisée.

Pouvez-vous tester à nouveau et me confirmer que ça fonctionne s.v.p?

**Comment by atifmohamed** [comment]
Effectivement, il semble que l'adresse IP de notre serveur est débloquée
Cet après midi, entre 13:00 et 13:05, j'ai effectué un essai, maintenant on a l'erreur suivante  : 

code_http: NotFound - Not Found     
http_client_erreur: "erreurs":null not present.         
retour: 
!-- saved from url=(0050)file://\\S20_intranet_01\InetMiroir\R_rrq\404.html -->

**Comment by loima02** [comment]
Bonjour,

De mon côté, je vois une erreur 404 (page not found).

Pourriez-vous valider votre fichier transmission `default.v0.transmission` et vous assurez qu'il est à jour sur notre serveur s.v.p?

Référence : [https://github.com/MTESSDev/FRW/blob/main/Documentation/fichiers-transmission.md](https://github.com/MTESSDev/FRW/blob/main/Documentation/fichiers-transmission.md)

Merci et bonne journée!

**Comment by loima02** [comment]
Bonjour,

Dans notre exemple complet (disponible ici : [https://github.com/MTESSDev/FRW/blob/main/Exemples/SystemeComplet/default.v0.transmission.yml](https://github.com/MTESSDev/FRW/blob/main/Exemples/SystemeComplet/default.v0.transmission.yml)), notre service de transmission de document retourne à FRW quelques informations incluant la chaîne de caractères `"erreurs":null`
```
    check_response:
      throw_exception_if_body_not_contains_all:
        - '"erreurs":null'
```

Si FRW ne détecte pas la chaîne de caractère indiquée dans le `check_response`, il lance une exception.

Libre à vous de configurer votre fichier de transmission pour valider un code de retour ou une chaîne de caractère.

Nous allons bonifier la documentation ici : [https://github.com/MTESSDev/FRW/blob/main/Documentation/fichiers-transmission.md](https://github.com/MTESSDev/FRW/blob/main/Documentation/fichiers-transmission.md)

Merci et bonne journée,

**Comment by anisite** [comment]
Pour ajouter un peu de précision, normalement le bloc ``check_response`` est facultatif, êtes-vous en mesure de valider que s'il est absent tout fonctionne ? Le comportement prévu est que par défaut tout code de retour ``Succès`` (20x) laissera terminer notre traitement sans problème et qu'en cas de code http de type erreur nous allons être alertés.

La partie ``check_response`` est un filet de sécurité de plus, qui permet de valider une information de plus dans votre réponse.

Actuellement, il n'est pas possible de valider un status code http précis, tous ceux qui sont dans la plage des 200 sont traités comme succès.

Derrière, c'est l'outil YamlHttpClient qui est utilisé les deux check disponibles sont ``throw_exception_if_body_contains_any`` et ``throw_exception_if_body_not_contains_all`` tel que démontré dans l'exemple :
https://github.com/anisite/YamlHttpClient#yaml-config-sample

N'hésitez pas à proposer des améliorations directement sur le github de YamlHttpClient ou autres, c'est nous qui sommes responsables de cet outil-là.

Information importante aussi, nous allons réessayer l'appel automatiquement en cas d'échec, mais pour éviter des aller-retour entres nous en cas de panne et l'accumulation de plusieurs demande à reprendre, il est préférable que votre traitement accepte les appels et qu'ils s'empilent de votre côté dans votre mécanisme de reprise.

Si ce n'est pas actuellement comme ça que vous l'avez implémenté, pas la fin du monde, mais pour le futur, nous allons demander à ce que la prise en charge soit à faible couplage, principalement pour limiter les interventions manuelles.

**Comment by atifmohamed** [comment]
Nous avons retiré la section check_response temporairement; Il semble que le problème de communication est à notre niveau, on travaille à le corriger, on revient vers vous au besoin. Merci pour votre disponibilité.

**Comment by atifmohamed** [comment]
Il semble que le problème de transmission de formulaire web est corrigé.
Ce matin entre 09:00 et 09:07, j'ai effectué un essai, un code 200 et un message "OK" de confirmation sont retournés.
Pourriez-vous nous confirmer que tout est correct dans votre côté ?

**Comment by loima02** [comment]
Bonjour!
Dans la période ciblée, je confirme que je vois des `CRE`, `REP` et `MAJ`.
Je ne vois pas de `TRANSMIS`.
Réf : [cycle-de-vie-etats.md](https://github.com/MTESSDev/FRW/blob/674fc7fc5ffb0a6d34bb1e39818ffbf4f2f0b937/Documentation/cycle-de-vie-etats.md?plain=1#L4)
Merci et bonne journée!

**Comment by atifmohamed** [comment]
Un autre essai est effectué vers 11:02.
Pourriez-vous vérifier les états générés par cet essais ?
--------------------------------------------
Le formulaire a bien été transmis.
Votre numéro de confirmation est le 50206311.
2023-06-07 11:02:07
-----------------------------------------------------

**Comment by loima02** [comment]
`CRE` -> `REP` -> `MAJ` -> `TRANSMIS`

**Comment by atifmohamed** [comment]
Merci pour votre belle collaboration, je ferme cette discussion avec l'état "Résolu"
